--- a/net/minecraft/world/inventory/ContainerLoom.java
+++ b/net/minecraft/world/inventory/ContainerLoom.java
@@ -24,8 +24,37 @@
 import net.minecraft.world.level.block.entity.EnumBannerPatternType;
 import net.minecraft.world.level.block.entity.TileEntityTypes;
 
+// CraftBukkit start
+import java.util.Objects;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.inventory.CraftInventoryLoom;
+import org.bukkit.craftbukkit.inventory.CraftInventoryView;
+import org.bukkit.craftbukkit.inventory.subcontainer.CraftChangeDetectingSubContainer;
+import org.bukkit.craftbukkit.inventory.view.CraftLoomView;
+import org.bukkit.entity.Player;
+// CraftBukkit end
+
 public class ContainerLoom extends Container {
 
+    // CraftBukkit start
+    private CraftLoomView bukkitEntity = null;
+    private Player player;
+
+    @Override
+    public CraftInventoryView getBukkitView() {
+        if (bukkitEntity != null) {
+            return bukkitEntity;
+        }
+
+        CraftInventoryLoom inventory = new CraftInventoryLoom(this.inputContainer, this.outputContainer);
+        bukkitEntity = new CraftLoomView(this.player, inventory, this);
+        return bukkitEntity;
+    }
+
+    public ContainerLoom(int i, PlayerInventory playerinventory, final ContainerAccess containeraccess) {
+        this(i, playerinventory, containeraccess, null, null);
+    }
+    // CraftBukkit end
     private static final int PATTERN_NOT_SET = -1;
     private static final int INV_SLOT_START = 4;
     private static final int INV_SLOT_END = 31;
@@ -47,27 +76,47 @@
         this(i, playerinventory, ContainerAccess.NULL);
     }
 
-    public ContainerLoom(int i, PlayerInventory playerinventory, final ContainerAccess containeraccess) {
+    // CraftBukkit start -> Add inputContainer and resultContainer parameters
+    public ContainerLoom(int i, PlayerInventory playerinventory, final ContainerAccess containeraccess, IInventory inputContainer, IInventory resultContainer) {
         super(Containers.LOOM, i);
         this.selectedBannerPatternIndex = ContainerProperty.standalone();
         this.selectablePatterns = List.of();
         this.slotUpdateListener = () -> {
         };
-        this.inputContainer = new InventorySubcontainer(3) {
+        // CraftBukkit start -> move inputContainer under if statement and set if not null
+        this.inputContainer = Objects.requireNonNullElseGet(inputContainer, () -> new InventorySubcontainer(3) {
             @Override
             public void setChanged() {
                 super.setChanged();
                 ContainerLoom.this.slotsChanged(this);
                 ContainerLoom.this.slotUpdateListener.run();
             }
-        };
-        this.outputContainer = new InventorySubcontainer(1) {
+
+            // CraftBukkit start
+            @Override
+            public Location getLocation() {
+                return containeraccess.getLocation();
+            }
+            // CraftBukkit end
+        });
+        if (this.inputContainer instanceof CraftChangeDetectingSubContainer changer) {
+            changer.setContainer(this::slotsChanged, containeraccess);
+        }
+        this.outputContainer = Objects.requireNonNullElseGet(resultContainer, () -> new InventorySubcontainer(1) {
             @Override
             public void setChanged() {
                 super.setChanged();
                 ContainerLoom.this.slotUpdateListener.run();
             }
-        };
+
+            // CraftBukkit start
+            @Override
+            public Location getLocation() {
+                return containeraccess.getLocation();
+            }
+            // CraftBukkit end
+        });
+        // CraftBukkit end
         this.access = containeraccess;
         this.bannerSlot = this.addSlot(new Slot(this.inputContainer, 0, 13, 26) {
             @Override
@@ -127,10 +176,12 @@
         }
 
         this.addDataSlot(this.selectedBannerPatternIndex);
+        player = (Player) playerinventory.player.getBukkitEntity(); // CraftBukkit
     }
 
     @Override
     public boolean stillValid(EntityHuman entityhuman) {
+        if (!this.checkReachable) return true; // CraftBukkit
         return stillValid(this.access, entityhuman, Blocks.LOOM);
     }
 
@@ -308,6 +359,11 @@
 
             if (nbttagcompound != null && nbttagcompound.contains("Patterns", 9)) {
                 nbttaglist = nbttagcompound.getList("Patterns", 10);
+                // CraftBukkit start
+                while (nbttaglist.size() > 20) {
+                    nbttaglist.remove(20);
+                }
+                // CraftBukkit end
             } else {
                 nbttaglist = new NBTTagList();
                 if (nbttagcompound == null) {
