--- a/net/minecraft/world/inventory/ICrafting.java
+++ b/net/minecraft/world/inventory/ICrafting.java
@@ -7,4 +7,10 @@
     void slotChanged(Container container, int i, ItemStack itemstack);
 
     void dataChanged(Container container, int i, int j);
+
+    // Craftbukkit start - Add slotChanged method that includes the previous ItemStack.
+    default void slotChanged(Container container, int i, ItemStack before, ItemStack after) {
+        slotChanged(container, i, after);
+    }
+    // Craftbukkit end
 }
