--- a/net/minecraft/world/level/block/BlockBell.java
+++ b/net/minecraft/world/level/block/BlockBell.java
@@ -138,6 +138,12 @@
                 enumdirection = (EnumDirection) world.getBlockState(blockposition).getValue(BlockBell.FACING);
             }
 
+            // CraftBukkit start
+            if (!org.bukkit.craftbukkit.event.CraftEventFactory.handleBellRingEvent(world, blockposition, enumdirection, entity)) {
+                return false;
+            }
+            // CraftBukkit end
+
             ((TileEntityBell) tileentity).onHit(enumdirection);
             world.playSound((EntityHuman) null, blockposition, SoundEffects.BELL_BLOCK, SoundCategory.BLOCKS, 2.0F, 1.0F);
             world.gameEvent(entity, GameEvent.BLOCK_CHANGE, blockposition);
