--- a/net/minecraft/world/level/chunk/storage/EntityStorage.java
+++ b/net/minecraft/world/level/chunk/storage/EntityStorage.java
@@ -33,7 +33,7 @@
     private static final Logger LOGGER = LogManager.getLogger();
     private static final String ENTITIES_TAG = "Entities";
     private static final String POSITION_TAG = "Position";
-    private final WorldServer level;
+    public final WorldServer level; // PAIL private -> public
     private final IOWorker worker;
     private final LongSet emptyChunks = new LongOpenHashSet();
     private final ThreadedMailbox<Runnable> entityDeserializerQueue;
@@ -51,8 +51,9 @@
         if (this.emptyChunks.contains(chunkcoordintpair.pair())) {
             return CompletableFuture.completedFuture(b(chunkcoordintpair));
         } else {
-            CompletableFuture completablefuture = this.worker.b(chunkcoordintpair);
-            Function function = (nbttagcompound) -> {
+            CompletableFuture<ChunkEntities<Entity>> future = new CompletableFuture<>(); // CraftBukkit - custom future
+            CompletableFuture<NBTTagCompound> completablefuture = this.worker.b(chunkcoordintpair); // CraftBukkit - decompile error
+            Function<NBTTagCompound, ChunkEntities<Entity>> function = (nbttagcompound) -> { // CraftBukkit - decompile error
                 if (nbttagcompound == null) {
                     this.emptyChunks.add(chunkcoordintpair.pair());
                     return b(chunkcoordintpair);
@@ -77,7 +78,31 @@
             ThreadedMailbox threadedmailbox = this.entityDeserializerQueue;
 
             Objects.requireNonNull(this.entityDeserializerQueue);
-            return completablefuture.thenApplyAsync(function, threadedmailbox::a);
+            // CraftBukkit start - Add a chunk load to load legacy entities. Since entities are normally only loaded when a chunk is loaded, it will only have a negative impact if a plugin manually loads entities
+            // return completablefuture.thenApplyAsync(function, threadedmailbox::a);
+            completablefuture.thenApply(nbttagcompound -> {
+                        CompletableFuture<NBTTagCompound> chunkFuture;
+                        if (nbttagcompound == null) {
+                            chunkFuture = level.getChunkProvider().b(chunkcoordintpair.x, chunkcoordintpair.z, net.minecraft.world.level.chunk.ChunkStatus.EMPTY, true).thenApplyAsync(either -> {
+                                net.minecraft.world.level.chunk.IChunkAccess chunkAccess = either.left().get();
+                                if (chunkAccess instanceof net.minecraft.world.level.chunk.Chunk) {
+                                    ((net.minecraft.world.level.chunk.Chunk) chunkAccess).addEntities();
+                                } else if (chunkAccess instanceof net.minecraft.world.level.chunk.ProtoChunkExtension) {
+                                    ((net.minecraft.world.level.chunk.ProtoChunkExtension) chunkAccess).v().addEntities();
+                                }
+                                return nbttagcompound;
+                            }, level.getMinecraftServer());
+                        } else {
+                            chunkFuture = CompletableFuture.completedFuture(nbttagcompound);
+                        }
+
+                        chunkFuture.thenApplyAsync(function, threadedmailbox::a).thenApply(future::complete);
+
+                        return null;
+                    });
+
+            return future;
+            // CraftBukkit end
         }
     }
 
