--- a/net/minecraft/world/level/storage/WorldNBTStorage.java
+++ b/net/minecraft/world/level/storage/WorldNBTStorage.java
@@ -12,6 +12,13 @@
 import net.minecraft.world.entity.player.EntityHuman;
 import org.slf4j.Logger;
 
+// CraftBukkit start
+import java.io.FileInputStream;
+import java.io.InputStream;
+import net.minecraft.server.level.EntityPlayer;
+import org.bukkit.craftbukkit.entity.CraftPlayer;
+// CraftBukkit end
+
 public class WorldNBTStorage {
 
     private static final Logger LOGGER = LogUtils.getLogger();
@@ -40,6 +47,22 @@
 
     }
 
+    // CraftBukkit start
+    public void save(NBTTagCompound compound, String id) {
+        try {
+            File file = File.createTempFile(id + "-", ".dat", this.playerDir);
+
+            NBTCompressedStreamTools.writeCompressed(compound, file);
+            File file1 = new File(this.playerDir, id + ".dat");
+            File file2 = new File(this.playerDir, id + ".dat_old");
+
+            SystemUtils.safeReplaceFile(file1, file, file2);
+        } catch (Exception exception) {
+            LOGGER.warn("Failed to save player data for " + id);
+        }
+    }
+    // CraftBukkit end
+
     @Nullable
     public NBTTagCompound load(EntityHuman entityhuman) {
         NBTTagCompound nbttagcompound = null;
@@ -55,6 +78,16 @@
         }
 
         if (nbttagcompound != null) {
+            // CraftBukkit start
+            if (entityhuman instanceof EntityPlayer) {
+                CraftPlayer player = (CraftPlayer) entityhuman.getBukkitEntity();
+                // Only update first played if it is older than the one we have
+                long modified = new File(this.playerDir, entityhuman.getUUID().toString() + ".dat").lastModified();
+                if (modified < player.getFirstPlayed()) {
+                    player.setFirstPlayed(modified);
+                }
+            }
+            // CraftBukkit end
             int i = GameProfileSerializer.getDataVersion(nbttagcompound, -1);
 
             entityhuman.load(DataFixTypes.PLAYER.updateToCurrentVersion(this.fixerUpper, nbttagcompound, i));
@@ -63,6 +96,22 @@
         return nbttagcompound;
     }
 
+    // CraftBukkit start
+    public NBTTagCompound getPlayerData(String s) {
+        try {
+            File file1 = new File(this.playerDir, s + ".dat");
+
+            if (file1.exists()) {
+                return NBTCompressedStreamTools.readCompressed((InputStream) (new FileInputStream(file1)));
+            }
+        } catch (Exception exception) {
+            LOGGER.warn("Failed to load player data for " + s);
+        }
+
+        return null;
+    }
+    // CraftBukkit end
+
     public String[] getSeenPlayers() {
         String[] astring = this.playerDir.list();
 
@@ -78,4 +127,10 @@
 
         return astring;
     }
+
+    // CraftBukkit start
+    public File getPlayerDir() {
+        return playerDir;
+    }
+    // CraftBukkit end
 }
