--- a/net/minecraft/world/level/levelgen/GeneratorSettingBase.java
+++ b/net/minecraft/world/level/levelgen/GeneratorSettingBase.java
@@ -45,7 +45,7 @@
     public static final ResourceKey<GeneratorSettingBase> CAVES = ResourceKey.create(IRegistry.NOISE_GENERATOR_SETTINGS_REGISTRY, new MinecraftKey("caves"));
     public static final ResourceKey<GeneratorSettingBase> FLOATING_ISLANDS = ResourceKey.create(IRegistry.NOISE_GENERATOR_SETTINGS_REGISTRY, new MinecraftKey("floating_islands"));
 
-    private GeneratorSettingBase(StructureSettings structuresettings, NoiseSettings noisesettings, IBlockData iblockdata, IBlockData iblockdata1, SurfaceRules.o surfacerules_o, int i, boolean flag, boolean flag1, boolean flag2, boolean flag3, boolean flag4, boolean flag5) {
+    public GeneratorSettingBase(StructureSettings structuresettings, NoiseSettings noisesettings, IBlockData iblockdata, IBlockData iblockdata1, SurfaceRules.o surfacerules_o, int i, boolean flag, boolean flag1, boolean flag2, boolean flag3, boolean flag4, boolean flag5) {
         this.structureSettings = structuresettings;
         this.noiseSettings = noisesettings;
         this.defaultBlock = iblockdata;
