--- a/net/minecraft/world/entity/monster/EntityDrowned.java
+++ b/net/minecraft/world/entity/monster/EntityDrowned.java
@@ -52,6 +52,11 @@
 import net.minecraft.world.level.pathfinder.PathType;
 import net.minecraft.world.phys.Vec3D;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityFireTridentEvent;
+// CraftBukkit end
+
 public class EntityDrowned extends EntityZombie implements IRangedEntity {
 
     public static final float NAUTILUS_SHELL_CHANCE = 0.03F;
@@ -239,15 +244,29 @@
 
     @Override
     public void performRangedAttack(EntityLiving entityliving, float f) {
-        EntityThrownTrident entitythrowntrident = new EntityThrownTrident(this.level, this, new ItemStack(Items.TRIDENT));
+        // CraftBukkit start -> pull out trident creation from inside the construction
+        ItemStack trident = this.getItemInHand(net.minecraft.world.entity.projectile.ProjectileHelper.getWeaponHoldingHand(this, Items.TRIDENT)); // CraftBukkit - Use Trident in hand like skeletons (SPIGOT-7025)
+        EntityThrownTrident entitythrowntrident = new EntityThrownTrident(this.level, this, trident);
+        // CraftBukkit end
         double d0 = entityliving.getX() - this.getX();
         double d1 = entityliving.getY(0.3333333333333333D) - entitythrowntrident.getY();
         double d2 = entityliving.getZ() - this.getZ();
         double d3 = Math.sqrt(d0 * d0 + d2 * d2);
 
         entitythrowntrident.shoot(d0, d1 + d3 * 0.20000000298023224D, d2, 1.6F, (float) (14 - this.level.getDifficulty().getId() * 4));
+        // CraftBukkit start
+        EntityFireTridentEvent entityFireTridentEvent = CraftEventFactory.callEntityFireTridentEvent(this, trident, entitythrowntrident, 0.8f);
+        if (entityFireTridentEvent.isCancelled()) {
+            entityFireTridentEvent.getProjectile().remove();
+            return;
+        }
+        // CraftBukkit end
         this.playSound(SoundEffects.DROWNED_SHOOT, 1.0F, 1.0F / (this.getRandom().nextFloat() * 0.4F + 0.8F));
-        this.level.addFreshEntity(entitythrowntrident);
+        // CraftBukkit start
+        if (entitythrowntrident.getBukkitEntity() == entityFireTridentEvent.getProjectile()) {
+            this.level.addFreshEntity(entitythrowntrident);
+        }
+        // CraftBukkit end
     }
 
     public void setSearchingForLand(boolean flag) {
