--- a/net/minecraft/util/INamable.java
+++ b/net/minecraft/util/INamable.java
@@ -27,7 +27,7 @@
     }
 
     static <E extends Enum<E> & INamable> INamable.a<E> fromEnumWithMapping(Supplier<E[]> supplier, Function<String, String> function) {
-        E[] ae = (Enum[]) supplier.get();
+        E[] ae = supplier.get(); // CraftBukkit - decompile error
 
         if (ae.length > 16) {
             Map<String, E> map = (Map) Arrays.stream(ae).collect(Collectors.toMap((oenum) -> {
@@ -37,7 +37,7 @@
             }));
 
             return new INamable.a<>(ae, (s) -> {
-                return s == null ? null : (Enum) map.get(s);
+                return s == null ? null : map.get(s); // CraftBukkit - decompile error
             });
         } else {
             return new INamable.a<>(ae, (s) -> {
@@ -45,7 +45,7 @@
                 int i = ae.length;
 
                 for (int j = 0; j < i; ++j) {
-                    E e0 = aenum[j];
+                    E e0 = (E) aenum[j]; // CraftBukkit - decompile error
 
                     if (((String) function.apply(((INamable) e0).getSerializedName())).equals(s)) {
                         return e0;
@@ -60,7 +60,7 @@
     static Keyable keys(final INamable[] ainamable) {
         return new Keyable() {
             public <T> Stream<T> keys(DynamicOps<T> dynamicops) {
-                Stream stream = Arrays.stream(ainamable).map(INamable::getSerializedName);
+                Stream<String> stream = Arrays.stream(ainamable).map(INamable::getSerializedName); // CraftBukkit - decompile error
 
                 Objects.requireNonNull(dynamicops);
                 return stream.map(dynamicops::createString);
@@ -68,8 +68,7 @@
         };
     }
 
-    /** @deprecated */
-    @Deprecated
+    // CraftBukkit - decompile error
     public static class a<E extends Enum<E> & INamable> implements Codec<E> {
 
         private final Codec<E> codec;
@@ -96,11 +95,11 @@
 
         @Nullable
         public E byName(@Nullable String s) {
-            return (Enum) this.resolver.apply(s);
+            return this.resolver.apply(s); // CraftBukkit - decompile error
         }
 
         public E byName(@Nullable String s, E e0) {
-            return (Enum) Objects.requireNonNullElse(this.byName(s), e0);
+            return Objects.requireNonNullElse(this.byName(s), e0); // CraftBukkit - decompile error
         }
     }
 }
