--- a/net/minecraft/world/level/block/grower/WorldGenTreeProvider.java
+++ b/net/minecraft/world/level/block/grower/WorldGenTreeProvider.java
@@ -14,6 +14,10 @@
 import net.minecraft.world.level.chunk.ChunkGenerator;
 import net.minecraft.world.level.levelgen.feature.WorldGenFeatureConfigured;
 
+// CraftBukkit start
+import net.minecraft.world.level.GeneratorAccessSeed;
+// CraftBukkit end
+
 public abstract class WorldGenTreeProvider {
 
     public WorldGenTreeProvider() {}
@@ -21,13 +25,13 @@
     @Nullable
     protected abstract ResourceKey<WorldGenFeatureConfigured<?, ?>> getConfiguredFeature(RandomSource randomsource, boolean flag);
 
-    public boolean growTree(WorldServer worldserver, ChunkGenerator chunkgenerator, BlockPosition blockposition, IBlockData iblockdata, RandomSource randomsource) {
+    public boolean growTree(GeneratorAccessSeed worldserver, ChunkGenerator chunkgenerator, BlockPosition blockposition, IBlockData iblockdata, RandomSource randomsource) { // CraftBukkit - Move to interface for delegation
         ResourceKey<WorldGenFeatureConfigured<?, ?>> resourcekey = this.getConfiguredFeature(randomsource, this.hasFlowers(worldserver, blockposition));
 
         if (resourcekey == null) {
             return false;
         } else {
-            Holder<WorldGenFeatureConfigured<?, ?>> holder = (Holder) worldserver.registryAccess().registryOrThrow(Registries.CONFIGURED_FEATURE).getHolder(resourcekey).orElse((Object) null);
+            Holder<WorldGenFeatureConfigured<?, ?>> holder = (Holder) worldserver.registryAccess().registryOrThrow(Registries.CONFIGURED_FEATURE).getHolder(resourcekey).orElse(null); // CraftBukkit - decompile error
 
             if (holder == null) {
                 return false;
@@ -38,7 +42,7 @@
                 worldserver.setBlock(blockposition, iblockdata1, 4);
                 if (worldgenfeatureconfigured.place(worldserver, chunkgenerator, randomsource, blockposition)) {
                     if (worldserver.getBlockState(blockposition) == iblockdata1) {
-                        worldserver.sendBlockUpdated(blockposition, iblockdata, iblockdata1, 2);
+                        //worldserver.sendBlockUpdated(blockposition, iblockdata, iblockdata1, 2); // CraftBukkit - Don't send update
                     }
 
                     return true;
@@ -65,4 +69,5 @@
 
         return true;
     }
+
 }
