--- a/net/minecraft/world/item/ItemSign.java
+++ b/net/minecraft/world/item/ItemSign.java
@@ -24,7 +24,13 @@
     @Override
     protected boolean updateCustomBlockEntityTag(BlockPosition blockposition, World world, @Nullable EntityHuman entityhuman, ItemStack itemstack, IBlockData iblockdata) {
         boolean flag = super.updateCustomBlockEntityTag(blockposition, world, entityhuman, itemstack, iblockdata);
+    // CraftBukkit start
+        return flag;
+    }
 
+    @Override
+    protected boolean updateCustomBlockEntityTagAfterPlaceEvent(BlockPosition blockposition, World world, @org.jetbrains.annotations.Nullable EntityHuman entityhuman, ItemStack itemstack, IBlockData iblockdata, boolean flag) {
+    // CraftBukkit end
         if (!world.isClientSide && !flag && entityhuman != null) {
             TileEntity tileentity = world.getBlockEntity(blockposition);
 
