--- a/net/minecraft/world/item/ItemCooldown.java
+++ b/net/minecraft/world/item/ItemCooldown.java
@@ -6,6 +6,12 @@
 import java.util.Map.Entry;
 import net.minecraft.util.MathHelper;
 
+// CraftBukkit start
+import net.minecraft.world.entity.player.EntityHuman;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.event.player.PlayerItemCooldownChangeEvent;
+// CraftBukkit end
+
 public class ItemCooldown {
 
     public final Map<Item, ItemCooldown.Info> cooldowns = Maps.newHashMap();
@@ -47,6 +53,16 @@
 
     }
 
+    // CraftBukkit start - Call PlayerItemCooldownChangeEvent
+    public void addCooldown(Item item, int i, EntityHuman player, PlayerItemCooldownChangeEvent.Reason reason) {
+        PlayerItemCooldownChangeEvent playeritemcooldownevent = new PlayerItemCooldownChangeEvent(player.getBukkitEntity(), CraftItemStack.asBukkitCopy(new ItemStack(item)).getType(), i, reason);
+        player.getServer().server.getPluginManager().callEvent(playeritemcooldownevent);
+        if (!playeritemcooldownevent.isCancelled()) {
+            addCooldown(item, playeritemcooldownevent.getCooldown());
+        }
+    }
+    // CraftBukkit end
+
     public void addCooldown(Item item, int i) {
         this.cooldowns.put(item, new ItemCooldown.Info(this.tickCount, this.tickCount + i));
         this.onCooldownStarted(item, i);
