--- a/net/minecraft/world/item/ItemTrident.java
+++ b/net/minecraft/world/item/ItemTrident.java
@@ -26,6 +26,11 @@
 import net.minecraft.world.level.block.state.IBlockData;
 import net.minecraft.world.phys.Vec3D;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityFireTridentEvent;
+// CraftBukkit end
+
 public class ItemTrident extends Item implements ItemVanishable {
 
     public static final int THROW_THRESHOLD_TIME = 10;
@@ -68,27 +73,60 @@
 
                 if (k <= 0 || entityhuman.isInWaterOrRain()) {
                     if (!world.isClientSide) {
+                        // CraftBukkit - moved down
+                        /*
                         itemstack.hurtAndBreak(1, entityhuman, (entityhuman1) -> {
                             entityhuman1.broadcastBreakEvent(entityliving.getUsedItemHand());
                         });
+                        */
                         if (k == 0) {
                             EntityThrownTrident entitythrowntrident = new EntityThrownTrident(world, entityhuman, itemstack);
+                            // CraftBukkit start
+                            EntityFireTridentEvent entityFireTridentEvent = CraftEventFactory.callEntityFireTridentEvent(entityhuman, itemstack, entitythrowntrident, 0.8f);
+                            if (entityFireTridentEvent.isCancelled()) {
+                                entityFireTridentEvent.getProjectile().remove();
+                                return;
+                            }
+                            // CraftBukkit end
 
                             entitythrowntrident.shootFromRotation(entityhuman, entityhuman.getXRot(), entityhuman.getYRot(), 0.0F, 2.5F + (float) k * 0.5F, 1.0F);
                             if (entityhuman.getAbilities().instabuild) {
                                 entitythrowntrident.pickup = EntityArrow.PickupStatus.CREATIVE_ONLY;
                             }
 
-                            world.addFreshEntity(entitythrowntrident);
+                            // CraftBukkit start
+                            if (entitythrowntrident.getBukkitEntity() == entityFireTridentEvent.getProjectile() && !world.addFreshEntity(entitythrowntrident)) {
+                                if (entityhuman instanceof net.minecraft.server.level.EntityPlayer) {
+                                    ((net.minecraft.server.level.EntityPlayer) entityhuman).getBukkitEntity().updateInventory();
+                                }
+                                return;
+                            }
+
+                            itemstack.hurtAndBreak(1, entityhuman, (entityhuman1) -> {
+                                entityhuman1.broadcastBreakEvent(entityliving.getUsedItemHand());
+                            });
+                            entitythrowntrident.tridentItem = itemstack.copy(); // SPIGOT-4511 update since damage call moved
+                            // CraftBukkit end
+
                             world.playSound((EntityHuman) null, (Entity) entitythrowntrident, SoundEffects.TRIDENT_THROW, SoundCategory.PLAYERS, 1.0F, 1.0F);
                             if (!entityhuman.getAbilities().instabuild) {
                                 entityhuman.getInventory().removeItem(itemstack);
                             }
+                            // CraftBukkit start - SPIGOT-5458 also need in this branch :(
+                        } else {
+                            itemstack.hurtAndBreak(1, entityhuman, (entityhuman1) -> {
+                                entityhuman1.broadcastBreakEvent(entityliving.getUsedItemHand());
+                            });
+                            // CraftBukkkit end
                         }
                     }
 
                     entityhuman.awardStat(StatisticList.ITEM_USED.get(this));
                     if (k > 0) {
+                        // CraftBukkit start
+                        org.bukkit.event.player.PlayerRiptideEvent event = new org.bukkit.event.player.PlayerRiptideEvent((org.bukkit.entity.Player) entityhuman.getBukkitEntity(), org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(itemstack));
+                        event.getPlayer().getServer().getPluginManager().callEvent(event);
+                        // CraftBukkit end
                         float f = entityhuman.getYRot();
                         float f1 = entityhuman.getXRot();
                         float f2 = -MathHelper.sin(f * 0.017453292F) * MathHelper.cos(f1 * 0.017453292F);
