--- a/net/minecraft/world/item/ItemChorusFruit.java
+++ b/net/minecraft/world/item/ItemChorusFruit.java
@@ -13,6 +13,8 @@
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.phys.Vec3D;
 
+import org.bukkit.event.player.PlayerItemCooldownChangeEvent; // CraftBukkit
+
 public class ItemChorusFruit extends Item {
 
     public ItemChorusFruit(Item.Info item_info) {
@@ -39,7 +41,16 @@
 
                 Vec3D vec3d = entityliving.position();
 
-                if (entityliving.randomTeleport(d3, d4, d5, true)) {
+                // CraftBukkit start - handle canceled status of teleport event
+                java.util.Optional<Boolean> status = entityliving.randomTeleport(d3, d4, d5, true, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.CHORUS_FRUIT);
+
+                if (!status.isPresent()) {
+                    // teleport event was canceled, no more tries
+                    break;
+                }
+
+                if (status.get()) {
+                    // CraftBukkit end
                     world.gameEvent(GameEvent.TELEPORT, vec3d, GameEvent.a.of((Entity) entityliving));
                     SoundEffect soundeffect = entityliving instanceof EntityFox ? SoundEffects.FOX_TELEPORT : SoundEffects.CHORUS_FRUIT_TELEPORT;
 
@@ -48,10 +59,11 @@
                     break;
                 }
             }
-
-            if (entityliving instanceof EntityHuman) {
-                ((EntityHuman) entityliving).getCooldowns().addCooldown(this, 20);
+            // CraftBukkit start - Call PlayerItemCooldownChangeEvent
+            if (entityliving instanceof EntityHuman entityHuman) {
+                entityHuman.getCooldowns().addCooldown(this, 20, entityHuman, PlayerItemCooldownChangeEvent.Reason.ITEM_USE);
             }
+            // CraftBukkit end
         }
 
         return itemstack1;
