--- a/net/minecraft/server/BlockJukeBox.java
+++ b/net/minecraft/server/BlockJukeBox.java
@@ -1,5 +1,12 @@
 package net.minecraft.server;
 
+//CraftBukkit start
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.event.block.JukeboxDropRecordEvent;
+import org.bukkit.event.player.PlayerJukeboxInsertRecordEvent;
+//CraftBukkit end
+
 import javax.annotation.Nullable;
 
 public class BlockJukeBox extends BlockTileEntity {
@@ -29,7 +36,11 @@
     @Override
     public EnumInteractionResult interact(IBlockData iblockdata, World world, BlockPosition blockposition, EntityHuman entityhuman, EnumHand enumhand, MovingObjectPositionBlock movingobjectpositionblock) {
         if ((Boolean) iblockdata.get(BlockJukeBox.HAS_RECORD)) {
-            this.dropRecord(world, blockposition);
+            //CraftBukkit start
+            if(!this.dropRecord(world, blockposition)){
+                return EnumInteractionResult.PASS;
+            }
+            //CraftBukkit end
             iblockdata = (IBlockData) iblockdata.set(BlockJukeBox.HAS_RECORD, false);
             world.setTypeAndData(blockposition, iblockdata, 2);
             return EnumInteractionResult.a(world.isClientSide);
@@ -38,16 +49,44 @@
         }
     }
 
+    //CraftBukkit add new method to accept a player input
+    public boolean insertRecordWithPlayer(GeneratorAccess generatoraccess, BlockPosition blockposition, IBlockData iblockdata, ItemStack itemstack, EntityHuman who){
+        TileEntity tileentity = generatoraccess.getTileEntity(blockposition);
+
+        if(tileentity instanceof TileEntityJukeBox){
+
+            if(who instanceof EntityPlayer){
+                PlayerJukeboxInsertRecordEvent playerJukeboxInsertRecordEvent = CraftEventFactory.callPlayerJukeboxInsertRecordEvent((EntityPlayer) who, (TileEntityJukeBox) tileentity, itemstack);
+                if(playerJukeboxInsertRecordEvent.isCancelled()){
+                    return false;
+                }
+
+                itemstack = CraftItemStack.asNMSCopy(playerJukeboxInsertRecordEvent.getRecord());
+            }
+
+            a(generatoraccess, blockposition, iblockdata, itemstack);
+        }
+
+        return true;
+    }
+    //CraftBukkit end
+
     public void a(GeneratorAccess generatoraccess, BlockPosition blockposition, IBlockData iblockdata, ItemStack itemstack) {
         TileEntity tileentity = generatoraccess.getTileEntity(blockposition);
 
         if (tileentity instanceof TileEntityJukeBox) {
-            ((TileEntityJukeBox) tileentity).setRecord(itemstack.cloneItemStack());
+            // CraftBukkit start - There can only be one
+            itemstack = itemstack.cloneItemStack();
+            if (!itemstack.isEmpty()) {
+                itemstack.setCount(1);
+            }
+            ((TileEntityJukeBox) tileentity).setRecord(itemstack);
+            // CraftBukkit end
             generatoraccess.setTypeAndData(blockposition, (IBlockData) iblockdata.set(BlockJukeBox.HAS_RECORD, true), 2);
         }
     }
 
-    public void dropRecord(World world, BlockPosition blockposition) {
+    public boolean dropRecord(World world, BlockPosition blockposition) { //CraftBukkit PAIL -> change return from void to boolean
         if (!world.isClientSide) {
             TileEntity tileentity = world.getTileEntity(blockposition);
 
@@ -56,6 +95,14 @@
                 ItemStack itemstack = tileentityjukebox.getRecord();
 
                 if (!itemstack.isEmpty()) {
+
+                    //CraftBukkit start
+                    JukeboxDropRecordEvent jukeboxDropRecordEvent = CraftEventFactory.callJukeboxDropRecordEvent(world, blockposition, itemstack);
+                    if(jukeboxDropRecordEvent.isCancelled()){
+                        return false;
+                    }
+                    itemstack = CraftItemStack.asNMSCopy(jukeboxDropRecordEvent.getRecord());
+                    //CraftBukkit end
                     world.triggerEffect(1010, blockposition, 0);
                     tileentityjukebox.clear();
                     float f = 0.7F;
@@ -67,9 +114,11 @@
 
                     entityitem.defaultPickupDelay();
                     world.addEntity(entityitem);
+                    return true;
                 }
             }
         }
+        return false; //CraftBukkit add return statement
     }
 
     @Override
