--- a/net/minecraft/server/ContainerShulkerBox.java
+++ b/net/minecraft/server/ContainerShulkerBox.java
@@ -1,11 +1,29 @@
 package net.minecraft.server;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.inventory.CraftInventoryView;
+// CraftBukkit end
+
 public class ContainerShulkerBox extends Container {
 
     private final IInventory a;
+    // CraftBukkit start
+    private CraftInventoryView bukkitEntity;
+    private PlayerInventory player;
+
+    @Override
+    public CraftInventoryView getBukkitView() {
+        if (bukkitEntity != null) {
+            return bukkitEntity;
+        }
+
+        return bukkitEntity = new CraftInventoryView(this.player.player.getBukkitEntity(), this.a.getBukkitInventory(), this);
+    }
+    // CraftBukkit end
 
     public ContainerShulkerBox(PlayerInventory playerinventory, IInventory iinventory, EntityHuman entityhuman) {
         this.a = iinventory;
+        this.player = playerinventory; // CraftBukkit - save player
         iinventory.startOpen(entityhuman);
         boolean flag = true;
         boolean flag1 = true;
