--- a/net/minecraft/server/DispenseBehaviorShulkerBox.java
+++ b/net/minecraft/server/DispenseBehaviorShulkerBox.java
@@ -1,5 +1,10 @@
 package net.minecraft.server;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.block.BlockDispenseEvent;
+// CraftBukkit end
+
 public class DispenseBehaviorShulkerBox extends DispenseBehaviorMaybe {
 
     public DispenseBehaviorShulkerBox() {}
@@ -14,6 +19,13 @@
             BlockPosition blockposition = isourceblock.getBlockPosition().shift(enumdirection);
             EnumDirection enumdirection1 = isourceblock.getWorld().isEmpty(blockposition.down()) ? enumdirection : EnumDirection.UP;
 
+            // CraftBukkit start
+            BlockDispenseEvent event = CraftEventFactory.callBlockDispenseEvent(isourceblock, itemstack, 1, new org.bukkit.util.Vector(blockposition.getX(), blockposition.getY(), blockposition.getZ()));
+            if (event == null) {
+                return itemstack;
+            }
+            // CraftBukkit end
+
             this.dispensed = ((ItemBlock) item).a((BlockActionContext) (new BlockActionContextDirectional(isourceblock.getWorld(), blockposition, enumdirection, itemstack, enumdirection1))) == EnumInteractionResult.SUCCESS;
         }
 
