--- a/net/minecraft/server/DispenseBehaviorBoat.java
+++ b/net/minecraft/server/DispenseBehaviorBoat.java
@@ -1,5 +1,10 @@
 package net.minecraft.server;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.block.BlockDispenseEvent;
+// CraftBukkit end
+
 public class DispenseBehaviorBoat extends DispenseBehaviorItem {
 
     private final DispenseBehaviorItem b = new DispenseBehaviorItem();
@@ -29,11 +34,20 @@
             d3 = 0.0D;
         }
 
-        EntityBoat entityboat = new EntityBoat(world, d0, d1 + d3, d2);
+        // EntityBoat entityboat = new EntityBoat(world, d0, d1 + d3, d2);
+        // CraftBukkit start
+        BlockDispenseEvent event = CraftEventFactory.callBlockDispenseEvent(isourceblock, itemstack, 1, new org.bukkit.util.Vector(d0, d1 + d3, d2));
+        if (event == null) {
+            return itemstack;
+        }
+
+        org.bukkit.util.Vector eventVelocity = event.getVelocity();
+        EntityBoat entityboat = new EntityBoat(world, eventVelocity.getX(), eventVelocity.getY(), eventVelocity.getZ());
+        // CraftBukkit end
 
         entityboat.setType(this.c);
         entityboat.yaw = enumdirection.l();
-        world.addEntity(entityboat);
+        if (!world.addEntity(entityboat)) return itemstack; // CraftBukkit
         itemstack.subtract(1);
         return itemstack;
     }
