--- a/net/minecraft/server/DragonControllerStrafe.java
+++ b/net/minecraft/server/DragonControllerStrafe.java
@@ -67,7 +67,13 @@
                         EntityDragonFireball entitydragonfireball = new EntityDragonFireball(this.a.world, this.a, d9, d10, d11);
 
                         entitydragonfireball.setPositionRotation(d6, d7, d8, 0.0F, 0.0F);
-                        this.a.world.addEntity(entitydragonfireball);
+                        // CraftBukkit start
+                        if (org.bukkit.craftbukkit.event.CraftEventFactory.callProjectileLaunchEvent(entitydragonfireball, ItemStack.a, null).isCancelled()) {
+                            entitydragonfireball.die();
+                            this.c = 0; // CraftBukkit - reset fireball charge
+                            return;
+                        }
+                        this.a.world.addEntity(entitydragonfireball, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.PROJECTILE); // CraftBukkit end
                         this.c = 0;
                         if (this.d != null) {
                             while (!this.d.b()) {
