--- a/net/minecraft/server/EntityLlama.java
+++ b/net/minecraft/server/EntityLlama.java
@@ -309,6 +309,8 @@
 
         entityllamaspit.shoot(d0, d1 + (double) f, d2, 1.5F, 10.0F);
         this.world.a((EntityHuman) null, this.locX, this.locY, this.locZ, SoundEffects.ENTITY_LLAMA_SPIT, this.bV(), 1.0F, 1.0F + (this.random.nextFloat() - this.random.nextFloat()) * 0.2F);
+        // CraftBukkit start - fire ProjectileLaunchEvent
+        if (org.bukkit.craftbukkit.event.CraftEventFactory.handleLaunchProjectileEvent(entityllamaspit, ItemStack.a, null, false)) return; // CraftBukkit - end
         this.world.addEntity(entityllamaspit);
         this.bP = true;
     }
