--- a/net/minecraft/server/EntityProjectile.java
+++ b/net/minecraft/server/EntityProjectile.java
@@ -3,6 +3,8 @@
 import java.util.List;
 import java.util.UUID;
 
+import org.bukkit.event.entity.ProjectileCollideWithEntityEvent; // CraftBukkit
+
 public abstract class EntityProjectile extends Entity implements IProjectile {
 
     private int blockX = -1;
@@ -26,6 +28,7 @@
     public EntityProjectile(World world, EntityLiving entityliving) {
         super(world);
         this.shooter = entityliving;
+        this.projectileSource = (org.bukkit.entity.LivingEntity) entityliving.getBukkitEntity(); // CraftBukkit
         this.setSize(0.25F, 0.25F);
         this.setPositionRotation(entityliving.locX, entityliving.locY + (double) entityliving.getHeadHeight(), entityliving.locZ, entityliving.yaw, entityliving.pitch);
         this.locX -= (double) (MathHelper.cos(this.yaw / 180.0F * 3.1415927F) * 0.16F);
@@ -126,6 +129,12 @@
                 Entity entity1 = (Entity) list.get(i);
 
                 if (entity1.ad() && (entity1 != entityliving || this.ar >= 5)) {
+                	//CraftBukkit start
+                	ProjectileCollideWithEntityEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callProjectileCollideWithEntityEvent(this, entity1);
+                	if (event.isCancelled()) {
+                		continue;
+                	}
+                	//CraftBukkit end
                     float f = 0.3F;
                     AxisAlignedBB axisalignedbb = entity1.getBoundingBox().grow((double) f, (double) f, (double) f);
                     MovingObjectPosition movingobjectposition1 = axisalignedbb.a(vec3d, vec3d1);
@@ -151,6 +160,11 @@
                 this.d(movingobjectposition.a());
             } else {
                 this.a(movingobjectposition);
+                // CraftBukkit start
+                if (this.dead) {
+                    org.bukkit.craftbukkit.event.CraftEventFactory.callProjectileHitEvent(this);
+                }
+                // CraftBukkit end
             }
         }
 
