--- a/net/minecraft/server/EntityBlaze.java
+++ b/net/minecraft/server/EntityBlaze.java
@@ -190,7 +190,11 @@
                             EntitySmallFireball entitysmallfireball = new EntitySmallFireball(this.a.world, this.a, d1 + this.a.getRandom().nextGaussian() * (double) f, d2, d3 + this.a.getRandom().nextGaussian() * (double) f);
 
                             entitysmallfireball.locY = this.a.locY + (double) (this.a.length / 2.0F) + 0.5D;
-                            this.a.world.addEntity(entitysmallfireball);
+                            // CraftBukkit start
+                            if (org.bukkit.craftbukkit.event.CraftEventFactory.callProjectileLaunchEvent(entitysmallfireball, ItemStack.a, null).isCancelled()) {
+                                return;
+                            }
+                            this.a.world.addEntity(entitysmallfireball, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.PROJECTILE); // CraftBukkit end - add SpawnReason
                         }
                     }
                 }
