--- a/net/minecraft/server/EntityLlama.java
+++ b/net/minecraft/server/EntityLlama.java
@@ -1,5 +1,7 @@
 package net.minecraft.server;
 
+import org.bukkit.event.entity.CreatureSpawnEvent;
+
 import java.util.Iterator;
 import java.util.function.Predicate;
 import javax.annotation.Nullable;
@@ -66,8 +68,8 @@
         this.goalSelector.a(6, new PathfinderGoalRandomStrollLand(this, 0.7D));
         this.goalSelector.a(7, new PathfinderGoalLookAtPlayer(this, EntityHuman.class, 6.0F));
         this.goalSelector.a(8, new PathfinderGoalRandomLookaround(this));
-        this.targetSelector.a(1, new EntityLlama.c(this));
-        this.targetSelector.a(2, new EntityLlama.a(this));
+        this.targetSelector.a(1, new c(this));
+        this.targetSelector.a(2, new a(this));
     }
 
     protected void initAttributes() {
@@ -176,7 +178,7 @@
             i = ((EntityLlama.b) object).a;
         } else {
             i = this.random.nextInt(4);
-            object = new EntityLlama.b(i, null);
+            object = new b(i, null);
         }
 
         this.setVariant(i);
@@ -309,7 +311,11 @@
 
         entityllamaspit.shoot(d0, d1 + (double) f, d2, 1.5F, 10.0F);
         this.world.a((EntityHuman) null, this.locX, this.locY, this.locZ, SoundEffects.ENTITY_LLAMA_SPIT, this.bV(), 1.0F, 1.0F + (this.random.nextFloat() - this.random.nextFloat()) * 0.2F);
-        this.world.addEntity(entityllamaspit);
+        // CraftBukkit - fire ProjectileLaunchEvent
+        if (org.bukkit.craftbukkit.event.CraftEventFactory.callProjectileLaunchEvent(entityllamaspit, ItemStack.a, null).isCancelled()) {
+            return;
+        }
+        this.world.addEntity(entityllamaspit, CreatureSpawnEvent.SpawnReason.NATURAL); // CraftBukkit end
         this.bP = true;
     }
 
