--- a/net/minecraft/server/EntityDrowned.java
+++ b/net/minecraft/server/EntityDrowned.java
@@ -185,6 +185,7 @@
         double d3 = (double) MathHelper.sqrt(d0 * d0 + d2 * d2);
 
         entitythrowntrident.shoot(d0, d1 + d3 * 0.20000000298023224D, d2, 1.6F, (float) (14 - this.world.getDifficulty().a() * 4));
+        if (org.bukkit.craftbukkit.event.CraftEventFactory.handleLaunchProjectileEvent(entitythrowntrident, entitythrowntrident.getItemStack(), EnumHand.MAIN_HAND, false)) return;
         this.a(SoundEffects.ENTITY_DROWNED_SHOOT, 1.0F, 1.0F / (this.getRandom().nextFloat() * 0.4F + 0.8F));
         this.world.addEntity(entitythrowntrident);
     }
@@ -205,7 +206,7 @@
             return this.a.f((EntityLiving) entityhuman);
         }
 
-        public boolean test(@Nullable Object object) {
+        public boolean test(@Nullable EntityHuman object) { // CraftBukkit - fix decompile error
             return this.a((EntityHuman) object);
         }
     }
