--- a/net/minecraft/server/EntityPig.java
+++ b/net/minecraft/server/EntityPig.java
@@ -2,6 +2,12 @@
 
 import javax.annotation.Nullable;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityTransformEvent;
+import org.bukkit.event.entity.EntityTransformEvent.TransformReason;
+// CraftBukkit end
+
 public class EntityPig extends EntityAnimal {
 
     private static final DataWatcherObject<Boolean> bC = DataWatcher.a(EntityPig.class, DataWatcherRegistry.i);
@@ -118,13 +124,14 @@
     }
 
     public void die(DamageSource damagesource) {
-        super.die(damagesource);
+        // super.die(damagesource); // CraftBukkit - Moved to end
         if (!this.world.isClientSide) {
             if (this.hasSaddle()) {
                 this.a((IMaterial) Items.SADDLE);
             }
 
         }
+        super.die(damagesource); // CraftBukkit - Moved from above
     }
 
     @Nullable
@@ -147,8 +154,18 @@
 
     public void onLightningStrike(EntityLightning entitylightning) {
         if (!this.world.isClientSide && !this.dead) {
+            // CraftBukkit start
+            EntityTransformEvent transform = CraftEventFactory.callEntityTransformEvent(this, TransformReason.LIGHTNING);
+            if (transform.isCancelled()) return;
+            // CraftBukkit end
             EntityPigZombie entitypigzombie = new EntityPigZombie(this.world);
 
+            // CraftBukkit start
+            if (CraftEventFactory.callPigZapEvent(this, entitylightning, entitypigzombie).isCancelled()) {
+                return;
+            }
+            // CraftBukkit end
+
             entitypigzombie.setSlot(EnumItemSlot.MAINHAND, new ItemStack(Items.GOLDEN_SWORD));
             entitypigzombie.setPositionRotation(this.locX, this.locY, this.locZ, this.yaw, this.pitch);
             entitypigzombie.setNoAI(this.isNoAI());
@@ -157,7 +174,8 @@
                 entitypigzombie.setCustomNameVisible(this.getCustomNameVisible());
             }
 
-            this.world.addEntity(entitypigzombie);
+            // CraftBukkit - added a reason for spawning this creature
+            this.world.addEntity(entitypigzombie, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.LIGHTNING);
             this.die();
         }
     }
