--- a/net/minecraft/server/EntityFallingBlock.java
+++ b/net/minecraft/server/EntityFallingBlock.java
@@ -5,6 +5,11 @@
 import java.util.Iterator;
 import javax.annotation.Nullable;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.block.AnvilDamageEvent;
+// CraftBukkit end
+
 public class EntityFallingBlock extends Entity {
 
     private IBlockData block;
@@ -65,7 +70,7 @@
 
             if (this.ticksLived++ == 0) {
                 blockposition = new BlockPosition(this);
-                if (this.world.getType(blockposition).getBlock() == block) {
+                if (this.world.getType(blockposition).getBlock() == block && !CraftEventFactory.callEntityChangeBlockEvent(this, blockposition.getX(), blockposition.getY(), blockposition.getZ(), Blocks.AIR, 0).isCancelled()) {
                     this.world.setAir(blockposition);
                 } else if (!this.world.isClientSide) {
                     this.die();
@@ -88,7 +93,7 @@
 
                     if (BlockFalling.i(this.world.getType(new BlockPosition(this.locX, this.locY - 0.009999999776482582D, this.locZ)))) {
                         this.onGround = false;
-                        return;
+                        // return; // CraftBukkit
                     }
 
                     this.motX *= 0.699999988079071D;
@@ -97,7 +102,13 @@
                     if (iblockdata.getBlock() != Blocks.PISTON_EXTENSION) {
                         this.die();
                         if (!this.f) {
-                            if (this.world.a(block, blockposition, true, EnumDirection.UP, (Entity) null, (ItemStack) null) && !BlockFalling.i(this.world.getType(blockposition.down())) && this.world.setTypeAndData(blockposition, this.block, 3)) {
+                            // CraftBukkit start
+                            if (this.world.a(block, blockposition, true, EnumDirection.UP, (Entity) null, (ItemStack) null) && !BlockFalling.i(this.world.getType(blockposition.down()))) {
+                                if (CraftEventFactory.callEntityChangeBlockEvent(this, blockposition.getX(), blockposition.getY(), blockposition.getZ(), this.block.getBlock(), this.block.getBlock().toLegacyData(this.block)).isCancelled()) {
+                                    return;
+                                }
+                                this.world.setTypeAndData(blockposition, this.block, 3);
+                                // CraftBukkit end
                                 if (block instanceof BlockFalling) {
                                     ((BlockFalling) block).a_(this.world, blockposition);
                                 }
@@ -154,18 +165,28 @@
                 while (iterator.hasNext()) {
                     Entity entity = (Entity) iterator.next();
 
+                    CraftEventFactory.entityDamage = this; // CraftBukkit
                     entity.damageEntity(damagesource, (float) Math.min(MathHelper.d((float) i * this.fallHurtAmount), this.fallHurtMax));
+                    CraftEventFactory.entityDamage = null; // CraftBukkit
                 }
 
                 if (flag && (double) this.random.nextFloat() < 0.05000000074505806D + (double) i * 0.05D) {
                     int j = ((Integer) this.block.get(BlockAnvil.DAMAGE)).intValue();
 
                     ++j;
-                    if (j > 2) {
-                        this.f = true;
-                    } else {
-                        this.block = this.block.set(BlockAnvil.DAMAGE, Integer.valueOf(j));
+                    // CaftBukkit stat
+                    AnvilDamageEvent event = CraftEventFactory.callAnvilDamageEvent(this.world, ((int) this.locX), ((int) this.locY), ((int) this.locZ), null, j, AnvilDamageEvent.DamageCause.FALLING);
+                    j = event.getNewState() == null ? 3 : event.getNewState().getData() / 4;
+                    if(!event.isCancelled()) {
+                        // CaftBukkit end
+                        if (j > 2) {
+                            this.f = true;
+                        } else {
+                            this.block = this.block.set(BlockAnvil.DAMAGE, Integer.valueOf(j));
+                        }
+                        // CaftBukkit stat
                     }
+                    // CaftBukkit end
                 }
             }
         }
