--- a/net/minecraft/server/EntityDolphin.java
+++ b/net/minecraft/server/EntityDolphin.java
@@ -7,6 +7,12 @@
 import java.util.Random;
 import java.util.function.Predicate;
 import javax.annotation.Nullable;
+//CraftBukkit start
+import org.bukkit.craftbukkit.entity.CraftItem;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.DolphinFollowEvent;
+import org.bukkit.event.entity.DolphinTossItemEvent;
+//CraftBukkit end
 
 public class EntityDolphin extends EntityWaterAnimal {
 
@@ -17,6 +23,7 @@
     public static final Predicate<EntityItem> b = (entityitem) -> {
         return !entityitem.p() && entityitem.isAlive() && entityitem.isInWater();
     };
+    public EntityHuman following; //CraftBukkit add variable outside of static class
 
     public EntityDolphin(EntityTypes<? extends EntityDolphin> entitytypes, World world) {
         super(entitytypes, world);
@@ -109,7 +116,7 @@
         this.goalSelector.a(8, new EntityDolphin.d());
         this.goalSelector.a(8, new PathfinderGoalFollowBoat(this));
         this.goalSelector.a(9, new PathfinderGoalAvoidTarget<>(this, EntityGuardian.class, 8.0F, 1.0D, 1.0D));
-        this.targetSelector.a(1, (new PathfinderGoalHurtByTarget(this, new Class[]{EntityGuardian.class})).a());
+        this.targetSelector.a(1, (new PathfinderGoalHurtByTarget(this, new Class[]{EntityGuardian.class})).a(new Class[0])); // CraftBukkit - decompile error
     }
 
     public static AttributeProvider.Builder eM() {
@@ -176,6 +183,12 @@
             ItemStack itemstack = entityitem.getItemStack();
 
             if (this.canPickup(itemstack)) {
+                // CraftBukkit start - call EntityPickupItemEvent
+                if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityPickupItemEvent(this, entityitem, 0, false).isCancelled()) {
+                    return;
+                }
+                itemstack = entityitem.getItemStack(); // update ItemStack from event
+                // CraftBukkit end
                 this.a(entityitem);
                 this.setSlot(EnumItemSlot.MAINHAND, itemstack);
                 this.dropChanceHand[EnumItemSlot.MAINHAND.b()] = 2.0F;
@@ -323,7 +336,7 @@
 
         @Override
         public boolean a() {
-            return this.a.gotFish() && this.a.getAirTicks() >= 100;
+            return this.a.gotFish() && this.a.getAirTicks() >= 100 && this.a.world.getWorld().canGenerateStructures(); // MC-151364, SPIGOT-5494: hangs if generate-structures=false
         }
 
         @Override
@@ -422,6 +435,16 @@
         @Override
         public boolean a() {
             this.c = this.a.world.a(EntityDolphin.bp, (EntityLiving) this.a);
+            //CraftBukkit start
+            if(this.c != null) {
+                DolphinFollowEvent dolphinFollowEvent = CraftEventFactory.callDolphinFollowEvent(a, c);
+                if (dolphinFollowEvent.isCancelled()) {
+                    this.c = null;
+                    return false;
+                }
+            }
+            a.following = this.c; //CraftBukkit clone the following into the dolphin object
+            //CraftBukkit end
             return this.c == null ? false : this.c.isSwimming() && this.a.getGoalTarget() != this.c;
         }
 
@@ -432,12 +455,14 @@
 
         @Override
         public void c() {
-            this.c.addEffect(new MobEffect(MobEffects.DOLPHINS_GRACE, 100));
+            this.c.addEffect(new MobEffect(MobEffects.DOLPHINS_GRACE, 100), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.DOLPHIN); // CraftBukkit
         }
 
         @Override
         public void d() {
+            CraftEventFactory.callDolphinUnfollowEvent(a, c); //CraftBukkit event call
             this.c = null;
+            a.following = this.c; //CraftBukkit clone the following into the dolphin object
             this.a.getNavigation().o();
         }
 
@@ -451,7 +476,7 @@
             }
 
             if (this.c.isSwimming() && this.c.world.random.nextInt(6) == 0) {
-                this.c.addEffect(new MobEffect(MobEffects.DOLPHINS_GRACE, 100));
+                this.c.addEffect(new MobEffect(MobEffects.DOLPHINS_GRACE, 100), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.DOLPHIN); // CraftBukkit
             }
 
         }
@@ -519,6 +544,10 @@
 
                 entityitem.setPickupDelay(40);
                 entityitem.setThrower(EntityDolphin.this.getUniqueID());
+                //CraftBukkit start
+                DolphinTossItemEvent dolphinTossItemEvent = CraftEventFactory.callDolphinTossItemEvent(EntityDolphin.this, entityitem);
+                entityitem = (EntityItem) ((CraftItem) dolphinTossItemEvent.getItemToToss()).getHandle();
+                //CraftBukkit end
                 float f = 0.3F;
                 float f1 = EntityDolphin.this.random.nextFloat() * 6.2831855F;
                 float f2 = 0.02F * EntityDolphin.this.random.nextFloat();
