--- a/net/minecraft/server/InventoryHorseChest.java
+++ b/net/minecraft/server/InventoryHorseChest.java
@@ -1,8 +1,23 @@
 package net.minecraft.server;
 
+import org.bukkit.craftbukkit.inventory.CraftInventoryAbstractHorse;
+
 public class InventoryHorseChest extends InventorySubcontainer {
 
-    public InventoryHorseChest(IChatBaseComponent ichatbasecomponent, int i) {
-        super(ichatbasecomponent, i);
+    // CraftBukkit start
+    public InventoryHorseChest(IChatBaseComponent ichatbasecomponent, int i, EntityHorseAbstract owner) {
+        super(ichatbasecomponent, i, (org.bukkit.entity.AbstractHorse) owner.getBukkitEntity());
+        // CraftBukkit end
+    }
+
+    // CraftBukkit start - return specialized inventory
+    @Override
+    public org.bukkit.inventory.Inventory getBukkitInventory() {
+        if (bukkitInventory == null) {
+            bukkitInventory = new CraftInventoryAbstractHorse(this);
+        }
+
+        return bukkitInventory;
     }
+    // CraftBukkit end
 }
