--- a/net/minecraft/server/ItemSplashPotion.java
+++ b/net/minecraft/server/ItemSplashPotion.java
@@ -8,17 +8,33 @@
 
     public InteractionResultWrapper<ItemStack> a(World world, EntityHuman entityhuman, EnumHand enumhand) {
         ItemStack itemstack = entityhuman.b(enumhand);
-        ItemStack itemstack1 = entityhuman.abilities.canInstantlyBuild ? itemstack.cloneItemStack() : itemstack.cloneAndSubtract(1);
+        // CraftBukkit - replace cloneAndSubtract with explicit calls so we can capture original itemstack in event
+        ItemStack itemstack1 = itemstack.cloneItemStack();
+        itemstack1.setCount(Math.min(1, itemstack.getCount())); // CraftBukkit - mimic cloneAndSubtract behavior
 
-        world.a((EntityHuman) null, entityhuman.locX, entityhuman.locY, entityhuman.locZ, SoundEffects.ENTITY_SPLASH_POTION_THROW, SoundCategory.PLAYERS, 0.5F, 0.4F / (ItemSplashPotion.k.nextFloat() * 0.4F + 0.8F));
+        // CraftBukkit moved down //world.a((EntityHuman) null, entityhuman.locX, entityhuman.locY, entityhuman.locZ, SoundEffects.ENTITY_SPLASH_POTION_THROW, SoundCategory.PLAYERS, 0.5F, 0.4F / (ItemSplashPotion.k.nextFloat() * 0.4F + 0.8F));
         if (!world.isClientSide) {
             EntityPotion entitypotion = new EntityPotion(world, entityhuman, itemstack1);
 
             entitypotion.a(entityhuman, entityhuman.pitch, entityhuman.yaw, -20.0F, 0.5F, 1.0F);
-            world.addEntity(entitypotion);
+            // CraftBukkit start - fire event
+            org.bukkit.event.entity.LaunchProjectileEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callLaunchProjectileEvent(entitypotion, itemstack, enumhand, false);
+            if (event.isCancelled() || !world.addEntity(entitypotion, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.THROWN)) {
+                entitypotion.die();
+                if (entityhuman instanceof EntityPlayer) {
+                    ((EntityPlayer)entityhuman).getBukkitEntity().updateInventory();
+                }
+                return new InteractionResultWrapper<>(EnumInteractionResult.FAIL, itemstack);
+            }
+            world.a((EntityHuman) null, entityhuman.locX, entityhuman.locY, entityhuman.locZ, SoundEffects.ENTITY_SPLASH_POTION_THROW, SoundCategory.PLAYERS, 0.5F, 0.4F / (ItemSplashPotion.k.nextFloat() * 0.4F + 0.8F));
+            entityhuman.b(StatisticList.ITEM_USED.b(this));
+            if (!entityhuman.abilities.canInstantlyBuild && event.shouldConsumeItem()) {
+                itemstack.subtract(1);
+            }
+            //world.addEntity(entitypotion); // CraftBukkit - moved up
         }
 
-        entityhuman.b(StatisticList.ITEM_USED.b(this));
-        return new InteractionResultWrapper(EnumInteractionResult.SUCCESS, itemstack);
+        //entityhuman.b(StatisticList.ITEM_USED.b(this)); // CraftBukkit end - moved up
+        return new InteractionResultWrapper<>(EnumInteractionResult.SUCCESS, itemstack);
     }
 }
