--- a/net/minecraft/server/ItemHoe.java
+++ b/net/minecraft/server/ItemHoe.java
@@ -24,7 +24,10 @@
 
             if (iblockdata != null) {
                 EntityHuman entityhuman = itemactioncontext.getEntity();
-
+                // CraftBukkit start - fire event
+                org.bukkit.event.entity.EntityChangeBlockEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callBlockChangeEvent(world, blockposition, iblockdata, entityhuman);
+                if (event.isCancelled()) return EnumInteractionResult.FAIL;
+                iblockdata = ((org.bukkit.craftbukkit.block.data.CraftBlockData)event.getBlockData()).getState(); // CraftBukkit end
                 world.a(entityhuman, blockposition, SoundEffects.ITEM_HOE_TILL, SoundCategory.BLOCKS, 1.0F, 1.0F);
                 if (!world.isClientSide) {
                     world.setTypeAndData(blockposition, iblockdata, 11);
