--- a/net/minecraft/server/ItemArmor.java
+++ b/net/minecraft/server/ItemArmor.java
@@ -4,6 +4,11 @@
 import java.util.List;
 import java.util.UUID;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.block.BlockDispenseArmorEvent;
+// CraftBukkit end
+
 public class ItemArmor extends Item {
 
     private static final UUID[] k = new UUID[]{UUID.fromString("845DB27C-C624-495F-8C9F-6020A9A58B6B"), UUID.fromString("D8499B04-0E66-4726-AB29-64469D734E0D"), UUID.fromString("9F3D476D-C118-4544-8365-64846904B48E"), UUID.fromString("2AD3F246-FEE1-4E67-B886-69FD380BB150")};
@@ -29,6 +34,12 @@
         } else {
             EntityLiving entityliving = (EntityLiving) list.get(0);
             EnumItemSlot enumitemslot = EntityInsentient.h(itemstack);
+            // CraftBukkit start
+            BlockDispenseArmorEvent event = CraftEventFactory.callBlockDispenseArmorEvent(isourceblock, itemstack, 1, entityliving);
+            if (event == null) {
+                return itemstack;
+            }
+            // CraftBukkit end
             ItemStack itemstack1 = itemstack.cloneAndSubtract(1);
 
             entityliving.setSlot(enumitemslot, itemstack1);
