--- a/net/minecraft/server/ItemBow.java
+++ b/net/minecraft/server/ItemBow.java
@@ -6,12 +6,16 @@
 
     public ItemBow(Item.Info item_info) {
         super(item_info);
+        // CraftBukkit start - obfuscator went a little crazy
+        /*
         this.a(new MinecraftKey("pull"), (itemstack, world, entityliving) -> {
             return entityliving == null ? 0.0F : (entityliving.dl().getItem() != Items.BOW ? 0.0F : (float) (itemstack.k() - entityliving.dm()) / 20.0F);
         });
         this.a(new MinecraftKey("pulling"), (itemstack, world, entityliving) -> {
             return entityliving != null && entityliving.isHandRaised() && entityliving.dl() == itemstack ? 1.0F : 0.0F;
         });
+        */
+        // CraftBukkit end
     }
 
     @Override
@@ -31,7 +35,7 @@
 
                 if ((double) f >= 0.1D) {
                     boolean flag1 = flag && itemstack1.getItem() == Items.ARROW;
-
+                    boolean willConsume = !flag1 && !entityhuman.abilities.canInstantlyBuild; // Craftbukkit
                     if (!world.isClientSide) {
                         ItemArrow itemarrow = (ItemArrow) ((ItemArrow) (itemstack1.getItem() instanceof ItemArrow ? itemstack1.getItem() : Items.ARROW));
                         EntityArrow entityarrow = itemarrow.a(world, itemstack1, (EntityLiving) entityhuman);
@@ -56,6 +60,20 @@
                         if (EnchantmentManager.getEnchantmentLevel(Enchantments.ARROW_FIRE, itemstack) > 0) {
                             entityarrow.setOnFire(100);
                         }
+                        // CraftBukkit start
+                        org.bukkit.event.entity.EntityShootBowEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callEntityShootBowEvent(entityhuman, itemstack, itemstack1, entityarrow, f, !flag1 && !entityhuman.abilities.canInstantlyBuild);
+                        willConsume = event.getWillConsumeArrow() == org.bukkit.event.Event.Result.ALLOW || (event.getWillConsumeArrow() == org.bukkit.event.Event.Result.DEFAULT && !event.isCancelled());
+                        if (event.isCancelled()) {
+                            event.getProjectile().remove();
+                            if (willConsume) {
+                                itemstack1.subtract(1);
+                                if (itemstack1.isEmpty()) {
+                                    entityhuman.inventory.f(itemstack1);
+                                }
+                            }
+                            return;
+                        }
+                        // CraftBukkit end
 
                         itemstack.damage(1, entityhuman, (entityhuman1) -> {
                             entityhuman1.d(entityhuman.getRaisedHand());
@@ -64,11 +82,23 @@
                             entityarrow.fromPlayer = EntityArrow.PickupStatus.CREATIVE_ONLY;
                         }
 
-                        world.addEntity(entityarrow);
+                        // CraftBukkit start
+                        if (event.getProjectile() == entityarrow.getBukkitEntity()) {
+                            if (!world.addEntity(entityarrow)) {
+                                if (entityhuman instanceof EntityPlayer) {
+                                    ((EntityPlayer) entityhuman).getBukkitEntity().updateInventory();
+                                }
+                                return;
+                            }
+                        }
+                        // CraftBukkit end
                     }
 
                     world.a((EntityHuman) null, entityhuman.locX, entityhuman.locY, entityhuman.locZ, SoundEffects.ENTITY_ARROW_SHOOT, SoundCategory.PLAYERS, 1.0F, 1.0F / (ItemBow.i.nextFloat() * 0.4F + 1.2F) + f * 0.5F);
-                    if (!flag1 && !entityhuman.abilities.canInstantlyBuild) {
+                    // CraftBukkit start
+                    if (willConsume) {
+                    // if (!flag1 && !entityhuman.abilities.canInstantlyBuild) {
+                    // CraftBukkit end
                         itemstack1.subtract(1);
                         if (itemstack1.isEmpty()) {
                             entityhuman.inventory.f(itemstack1);
