--- a/net/minecraft/server/ItemMinecart.java
+++ b/net/minecraft/server/ItemMinecart.java
@@ -1,5 +1,10 @@
 package net.minecraft.server;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.block.BlockDispenseEvent;
+// CraftBukkit end
+
 public class ItemMinecart extends Item {
 
     private static final IDispenseBehavior a = new DispenseBehaviorItem() {
@@ -38,13 +43,22 @@
                 }
             }
 
-            EntityMinecartAbstract entityminecartabstract = EntityMinecartAbstract.a(world, d0, d1 + d3, d2, ((ItemMinecart) itemstack.getItem()).b);
+            // CraftBukkit start
+            // EntityMinecartAbstract entityminecartabstract = EntityMinecartAbstract.a(world, d0, d1 + d3, d2, ((ItemMinecart) itemstack.getItem()).b);
+            BlockDispenseEvent event = CraftEventFactory.callBlockDispenseEvent(isourceblock, itemstack, 1, new org.bukkit.util.Vector(d0, d1 + d3, d2));
+            if (event == null) {
+                return itemstack;
+            }
+
+            org.bukkit.util.Vector eventVelocity = event.getVelocity();
+            EntityMinecartAbstract entityminecartabstract = EntityMinecartAbstract.a(world, eventVelocity.getX(), eventVelocity.getY(), eventVelocity.getZ(), ((ItemMinecart) itemstack.getItem()).b);
+            // CraftBukkit end
 
             if (itemstack.hasName()) {
                 entityminecartabstract.setCustomName(itemstack.getName());
             }
 
-            world.addEntity(entityminecartabstract);
+            if (!world.addEntity(entityminecartabstract)) return itemstack; // CraftBukkit
             itemstack.subtract(1);
             return itemstack;
         }
@@ -87,7 +101,12 @@
                     entityminecartabstract.setCustomName(itemstack.getName());
                 }
 
-                world.addEntity(entityminecartabstract);
+                // CraftBukkit start
+                if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityPlaceEvent(itemactioncontext, entityminecartabstract).isCancelled()) {
+                    return EnumInteractionResult.FAIL;
+                }
+                // CraftBukkit end
+                if (!world.addEntity(entityminecartabstract)) return EnumInteractionResult.PASS; // CraftBukkit
             }
 
             itemstack.subtract(1);
