--- a/net/minecraft/server/ItemLingeringPotion.java
+++ b/net/minecraft/server/ItemLingeringPotion.java
@@ -8,17 +8,29 @@
 
     public InteractionResultWrapper<ItemStack> a(World world, EntityHuman entityhuman, EnumHand enumhand) {
         ItemStack itemstack = entityhuman.b(enumhand);
-        ItemStack itemstack1 = entityhuman.abilities.canInstantlyBuild ? itemstack.cloneItemStack() : itemstack.cloneAndSubtract(1);
+        // CraftBukkit start
+        ItemStack itemstack1 = entityhuman.abilities.canInstantlyBuild ? itemstack.cloneItemStack() : itemstack.cloneAndSubtract(1); // CraftBukkit - use our logic
 
-        world.a((EntityHuman) null, entityhuman.locX, entityhuman.locY, entityhuman.locZ, SoundEffects.ENTITY_LINGERING_POTION_THROW, SoundCategory.NEUTRAL, 0.5F, 0.4F / (ItemLingeringPotion.k.nextFloat() * 0.4F + 0.8F));
+        // CraftBukkit - moved down -> //world.a((EntityHuman) null, entityhuman.locX, entityhuman.locY, entityhuman.locZ, SoundEffects.ENTITY_LINGERING_POTION_THROW, SoundCategory.NEUTRAL, 0.5F, 0.4F / (ItemLingeringPotion.k.nextFloat() * 0.4F + 0.8F));
         if (!world.isClientSide) {
             EntityPotion entitypotion = new EntityPotion(world, entityhuman, itemstack1);
 
             entitypotion.a(entityhuman, entityhuman.pitch, entityhuman.yaw, -20.0F, 0.5F, 1.0F);
-            world.addEntity(entitypotion);
+            org.bukkit.event.entity.LaunchProjectileEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callLaunchProjectileEvent(entitypotion, itemstack, enumhand, false);
+            if (event.isCancelled() || !world.addEntity(entitypotion, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.THROWN)) {
+                itemstack.add(1);
+                entitypotion.die();
+                return new InteractionResultWrapper<>(EnumInteractionResult.FAIL, itemstack);
+            }
+            world.a((EntityHuman) null, entityhuman.locX, entityhuman.locY, entityhuman.locZ, SoundEffects.ENTITY_LINGERING_POTION_THROW, SoundCategory.NEUTRAL, 0.5F, 0.4F / (ItemLingeringPotion.k.nextFloat() * 0.4F + 0.8F));
+            entityhuman.b(StatisticList.ITEM_USED.b(this));
+            if (!entityhuman.abilities.canInstantlyBuild && !event.shouldConsumeItem()) {
+                itemstack.add(1); // add 1 back to original itemstack because cloneAndSubtract removes 1
+            }
+            //world.addEntity(entitypotion); // CraftBukkit - moved up
         }
 
-        entityhuman.b(StatisticList.ITEM_USED.b(this));
-        return new InteractionResultWrapper(EnumInteractionResult.SUCCESS, itemstack);
+        //entityhuman.b(StatisticList.ITEM_USED.b(this)); // CraftBukkit end - moved up
+        return new InteractionResultWrapper<>(EnumInteractionResult.SUCCESS, itemstack);
     }
 }
