--- a/net/minecraft/server/RecipeItemStack.java
+++ b/net/minecraft/server/RecipeItemStack.java
@@ -5,27 +5,37 @@
 import it.unimi.dsi.fastutil.ints.IntComparators;
 import it.unimi.dsi.fastutil.ints.IntList;
 import javax.annotation.Nullable;
+import java.util.logging.Logger;
 
 public class RecipeItemStack implements Predicate<ItemStack> {
 
-    public static final RecipeItemStack a = new RecipeItemStack(new ItemStack[0], null) {
+    public static final RecipeItemStack a = new RecipeItemStack(new ItemStack[0]) { // CraftBukkit fix decompile error
         public boolean a(@Nullable ItemStack itemstack) {
             return itemstack.isEmpty();
         }
 
-        public boolean apply(@Nullable Object object) {
+        public boolean apply(@Nullable ItemStack object) { // CraftBukkit fix decompile error
             return this.a((ItemStack) object);
         }
     };
     public final ItemStack[] choices;
+    public boolean exactMatch = false;
     private IntList c;
 
-    private RecipeItemStack(ItemStack... aitemstack) {
+    public RecipeItemStack(ItemStack... aitemstack) { // CraftBukkit private -> public
         this.choices = aitemstack;
+        this.exactMatch = false;
     }
 
+    // CraftBukkit start - add constructor
+    public RecipeItemStack(boolean exactMatch, ItemStack... itemstacks) {
+        this(itemstacks);
+        this.exactMatch = exactMatch;
+    }
+    // CraftBukkit end
+
     public boolean a(@Nullable ItemStack itemstack) {
-        if (itemstack == null) {
+        if (itemstack == null || itemstack.isEmpty()) { // CraftBukkit add isEmpty check
             return false;
         } else {
             ItemStack[] aitemstack = this.choices;
@@ -38,7 +48,7 @@
                     int k = itemstack1.getData();
 
                     if (k == 32767 || k == itemstack.getData()) {
-                        return true;
+                        return !exactMatch || ItemStack.equals(itemstack, itemstack1); // CraftBukkit - enable exact nbt matching in recipes
                     }
                 }
             }
@@ -96,11 +106,12 @@
         return RecipeItemStack.a;
     }
 
-    public boolean apply(@Nullable Object object) {
-        return this.a((ItemStack) object);
+
+    public boolean apply(@Nullable ItemStack object) { // CraftBukkit fix decompile error
+        return this.a(object);
     }
 
-    RecipeItemStack(ItemStack[] aitemstack, Object object) {
+    private RecipeItemStack(ItemStack[] aitemstack, Object object) { // CraftBukkit make constructor private
         this(aitemstack);
     }
 }
