--- a/net/minecraft/server/RemoteControlCommandListener.java
+++ b/net/minecraft/server/RemoteControlCommandListener.java
@@ -4,6 +4,7 @@
 
     private final StringBuffer a = new StringBuffer();
     private final MinecraftServer b;
+    protected org.bukkit.craftbukkit.command.CraftRemoteConsoleCommandSender sender = null; // CraftBukkit - add sender
 
     public RemoteControlCommandListener(MinecraftServer minecraftserver) {
         this.b = minecraftserver;
@@ -21,6 +22,12 @@
         return "Rcon";
     }
 
+    // CraftBukkit start - Send a String
+    public void sendMessage(String message) {
+        this.a.append(message);
+    }
+    // CraftBukkit end
+
     public void sendMessage(IChatBaseComponent ichatbasecomponent) {
         this.a.append(ichatbasecomponent.toPlainText());
     }
