--- a/net/minecraft/server/TileEntityContainer.java
+++ b/net/minecraft/server/TileEntityContainer.java
@@ -1,5 +1,10 @@
 package net.minecraft.server;
 
+//CraftBukkit start
+import org.bukkit.event.inventory.InventoryType;
+import org.bukkit.inventory.InventoryHolder;
+//CraftBukkit end
+
 public abstract class TileEntityContainer extends TileEntity implements ITileInventory {
 
     private ChestLock a;
@@ -34,4 +39,28 @@
     public void setLock(ChestLock chestlock) {
         this.a = chestlock;
     }
+
+    // CraftBukkit start
+    private boolean overrideInventoryHolder;
+    private org.bukkit.inventory.InventoryHolder owner;
+    protected void setOwner(org.bukkit.inventory.InventoryHolder owner) {
+        this.owner = owner;
+        this.overrideInventoryHolder = true;
+    }
+
+    public boolean hasOverriddenInventoryHolder() {
+        return overrideInventoryHolder;
+    }
+
+    @Override
+    public InventoryHolder getOwner() {
+        return hasOverriddenInventoryHolder() ? owner : super.getOwner();
+    }
+
+    @Override
+    public org.bukkit.Location getLocation() {
+        if (world == null) return null;
+        return new org.bukkit.Location(world.getWorld(), position.getX(), position.getY(), position.getZ());
+    }
+    // CraftBukkit end
 }
